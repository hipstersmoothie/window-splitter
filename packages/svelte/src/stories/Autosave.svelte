<script module>
  import PanelGroup from "../lib/PanelGroup.svelte";
  import Panel from "../lib/Panel.svelte";
  import PanelResizer from "../lib/PanelResizer.svelte";

  let panelGroup;

  /** @type {import("@window-splitter/interface").PanelGroupHandle} */
  export const handle = {
    getId: () => "panel-group",
    getTemplate: () => panelGroup.getTemplate(),
    getState: () => panelGroup.getState(),
    getPixelSizes: () => panelGroup.getPixelSizes(),
    getPercentageSizes: () => panelGroup.getPercentageSizes(),
    setSizes: (sizes) => panelGroup.setSizes(sizes),
  };
</script>

<PanelGroup
  bind:this={panelGroup}
  class="panel-group"
  autosaveId="autosave-example-svelte"
>
  <Panel id="1" class="panel">1</Panel>
  <PanelResizer id="resizer" class="panel-resizer" size="10px" />
  <Panel id="2" class="panel">2</Panel>
</PanelGroup>

<style>
  :global(.panel-group) {
    border: 1px solid rgba(0, 0, 0, 0.3);
    background: rgba(0, 0, 0, 0.1);
    border-radius: 12px;
    box-sizing: border-box;
    width: 500px;
  }

  :global(.panel) {
    height: 100%;
    width: 100%;
    padding: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
    box-sizing: border-box;
  }

  :global(.panel-resizer) {
    background: red;
  }
</style>
