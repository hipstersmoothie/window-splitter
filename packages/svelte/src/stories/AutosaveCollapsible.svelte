<script>
  import PanelGroup from "../lib/PanelGroup.svelte";
  import Panel from "../lib/Panel.svelte";
  import PanelResizer from "../lib/PanelResizer.svelte";

  const { onCollapseChange, handle: handleProp } = $props();

  let panelGroup;

  if (handleProp) {
    /** @type {import("@window-splitter/interface").PanelGroupHandle} */
    handleProp.current = {
      getId: () => "panel-group",
      getTemplate: () => panelGroup.getTemplate(),
      getState: () => panelGroup.getState(),
      getPixelSizes: () => panelGroup.getPixelSizes(),
      getPercentageSizes: () => panelGroup.getPercentageSizes(),
      setSizes: (sizes) => panelGroup.setSizes(sizes),
    };
  }
</script>

<PanelGroup
  bind:this={panelGroup}
  class="panel-group"
  autosaveId="autosave-collapsible-svelte"
>
  <Panel
    id="1"
    collapsible
    collapsedSize="100px"
    {onCollapseChange}
    min="140px"
    class="panel"
  >
    Collapsible
  </Panel>
  <PanelResizer id="resizer" class="panel-resizer" size="10px" />
  <Panel id="2" class="panel">Panel 2</Panel>
</PanelGroup>

<style>
  :global(.panel-group) {
    border: 1px solid rgba(0, 0, 0, 0.3);
    background: rgba(0, 0, 0, 0.1);
    border-radius: 12px;
    box-sizing: border-box;
    width: 500px;
  }

  :global(.panel) {
    height: 100%;
    width: 100%;
    padding: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
    box-sizing: border-box;
  }

  :global(.panel-resizer) {
    background: red;
  }
</style>
