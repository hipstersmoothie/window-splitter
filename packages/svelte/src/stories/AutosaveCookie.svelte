<script>
  import PanelGroup from "../lib/PanelGroup.svelte";
  import Panel from "../lib/Panel.svelte";
  import PanelResizer from "../lib/PanelResizer.svelte";

  const { snapshot, handle } = $props();

  let panelGroup;

  /** @type {import("@window-splitter/interface").PanelGroupHandle} */
  handle.current = {
    getId: () => "panel-group",
    getTemplate: () => panelGroup.getTemplate(),
    getState: () => panelGroup.getState(),
    getPixelSizes: () => panelGroup.getPixelSizes(),
    getPercentageSizes: () => panelGroup.getPercentageSizes(),
    setSizes: (sizes) => panelGroup.setSizes(sizes),
  };
</script>

<PanelGroup
  bind:this={panelGroup}
  class="panel-group h-200 w-502"
  autosaveId="autosave-cookie-svelte"
  autosaveStrategy="cookie"
  {snapshot}
>
  <Panel id="1" class="panel">Panel 1</Panel>
  <PanelResizer id="resizer" class="panel-resizer" size="10px" />
  <Panel id="2" class="panel">Panel 2</Panel>
</PanelGroup>

<style>
  :global(.h-200) {
    height: 200px;
  }

  :global(.w-502) {
    width: 502px !important;
  }

  :global(.panel-group) {
    border: 1px solid rgba(0, 0, 0, 0.3);
    background: rgba(0, 0, 0, 0.1);
    border-radius: 12px;
    box-sizing: border-box;
  }

  :global(.panel) {
    height: 100%;
    width: 100%;
    padding: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
    box-sizing: border-box;
  }

  :global(.panel-resizer) {
    background: red;
  }
</style>
