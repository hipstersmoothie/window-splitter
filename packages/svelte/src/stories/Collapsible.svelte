<script module>
  import PanelGroup from "../lib/PanelGroup.svelte";
  import Panel from "../lib/Panel.svelte";
  import PanelResizer from "../lib/PanelResizer.svelte";

  /** @type {import("@window-splitter/interface").PanelGroupHandle} */
  let panelGroup;
  /** @type {import("@window-splitter/interface").PanelHandle} */
  let leftHandleRef;
  /** @type {import("@window-splitter/interface").PanelHandle} */
  let rightHandleRef;

  /** @type {import("@window-splitter/interface").PanelGroupHandle} */
  export const handle = {
    getId: () => "panel-group",
    getTemplate: () => panelGroup.getTemplate(),
    getState: () => panelGroup.getState(),
    getPixelSizes: () => panelGroup.getPixelSizes(),
    getPercentageSizes: () => panelGroup.getPercentageSizes(),
    setSizes: (sizes) => panelGroup.setSizes(sizes),
  };

  /** @type {import("@window-splitter/interface").PanelHandle} */
  export const leftHandle = {
    collapse: () => leftHandleRef.collapse(),
    isCollapsed: () => leftHandleRef.isCollapsed(),
    expand: () => leftHandleRef.expand(),
    isExpanded: () => leftHandleRef.isExpanded(),
    getId: () => leftHandleRef.getId(),
    getPixelSize: () => leftHandleRef.getPixelSize(),
    getPercentageSize: () => leftHandleRef.getPercentageSize(),
    setSize: (size) => leftHandleRef.setSize(size),
  };

  /** @type {import("@window-splitter/interface").PanelHandle} */
  export const rightHandle = {
    collapse: () => rightHandleRef.collapse(),
    isCollapsed: () => rightHandleRef.isCollapsed(),
    expand: () => rightHandleRef.expand(),
    isExpanded: () => rightHandleRef.isExpanded(),
    getId: () => rightHandleRef.getId(),
    getPixelSize: () => rightHandleRef.getPixelSize(),
    getPercentageSize: () => rightHandleRef.getPercentageSize(),
  };

  let collapsed = $state(true);
  const onCollapseChange = (newCollapsed) => {
    collapsed = newCollapsed;
  };
</script>

<PanelGroup bind:this={panelGroup} class="panel-group">
  <Panel
    bind:this={leftHandleRef}
    class="collapsible-example-panel-1 panel"
    min="100px"
    collapsible
    collapsedSize="60px"
    collapseAnimation={{ easing: "bounce", duration: 1000 }}
    onCollapseChange={(isCollapsed) => {
      // eslint-disable-next-line no-console
      console.log("COLLAPSE PASSIVE", isCollapsed);
    }}
  >
    1
  </Panel>
  <PanelResizer class="panel-resizer" id="resizer-1" size="10px" />
  <Panel min="100px">2</Panel>
  <PanelResizer class="panel-resizer" id="resizer-2" size="10px" />
  <Panel
    bind:this={rightHandleRef}
    class="collapsible-example-panel-2 panel"
    min="100px"
    collapsible
    collapsedSize="60px"
    collapseAnimation={{ easing: "bounce", duration: 1000 }}
    {collapsed}
    {onCollapseChange}
  >
    3
  </Panel>
</PanelGroup>

<style>
  :global(.collapsible-example-panel-1) {
    border: 10px solid green;
    box-sizing: border-box;
  }

  :global(.panel-group) {
    border: 1px solid rgba(0, 0, 0, 0.3);
    background: rgba(0, 0, 0, 0.1);
    border-radius: 12px;
    box-sizing: border-box;
    width: 500px;
  }

  :global(.panel) {
    height: 100%;
    width: 100%;
    padding: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
    box-sizing: border-box;
  }

  :global(.panel-resizer) {
    background: red;
  }
</style>
