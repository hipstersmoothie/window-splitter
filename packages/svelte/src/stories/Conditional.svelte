<script module>
  import PanelGroup from "../PanelGroup.svelte";
  import Panel from "../Panel.svelte";
  import PanelResizer from "../PanelResizer.svelte";

  let panelGroup;

  /** @type {import("@window-splitter/interface").PanelGroupHandle} */
  export const handle = {
    getId: () => "panel-group",
    getTemplate: () => panelGroup.getTemplate(),
    getState: () => panelGroup.getState(),
    getPixelSizes: () => panelGroup.getPixelSizes(),
    getPercentageSizes: () => panelGroup.getPercentageSizes(),
    setSizes: (sizes) => panelGroup.setSizes(sizes),
  };

  let isExpanded = $state(false);
  const toggleExpanded = () => {
    isExpanded = !isExpanded;
  };
</script>

<PanelGroup bind:this={panelGroup} class="panel-group">
  <Panel
    id="panel-1"
    min="100px"
    collapsible
    collapsedSize="60px"
    class="panel"
  >
    1
  </Panel>
  <PanelResizer id="handle-1" class="panel-resizer" size="10px" />
  <Panel id="panel-2" min="100px" class="panel">2</Panel>

  {#if isExpanded}
    <PanelResizer id="handle-2" class="panel-resizer" size="10px" />
    <Panel id="panel-3" min="100px" class="panel">
      3
      <button type="button" onclick={toggleExpanded}>Close</button>
    </Panel>
  {/if}
</PanelGroup>
<button type="button" onclick={toggleExpanded}>Expand</button>

<style>
  :global(.panel-group) {
    border: 1px solid rgba(0, 0, 0, 0.3);
    background: rgba(0, 0, 0, 0.1);
    border-radius: 12px;
    box-sizing: border-box;
    width: 500px;
  }

  :global(.panel) {
    height: 100%;
    width: 100%;
    padding: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
    box-sizing: border-box;
  }

  :global(.panel-resizer) {
    background: red;
  }
</style>
