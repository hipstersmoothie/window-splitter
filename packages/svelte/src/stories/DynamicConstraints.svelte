<script module>
  import PanelGroup from "../PanelGroup.svelte";
  import Panel from "../Panel.svelte";
  import PanelResizer from "../PanelResizer.svelte";

  let panelGroup;

  /** @type {import("@window-splitter/interface").PanelGroupHandle} */
  export const handle = {
    getId: () => "panel-group",
    getTemplate: () => panelGroup.getTemplate(),
    getState: () => panelGroup.getState(),
    getPixelSizes: () => panelGroup.getPixelSizes(),
    getPercentageSizes: () => panelGroup.getPercentageSizes(),
    setSizes: (sizes) => panelGroup.setSizes(sizes),
  };

  let dynamicConstraints = $state(false);
  const toggleDynamicConstraints = () => {
    dynamicConstraints = !dynamicConstraints;
  };
</script>

<PanelGroup class="panel-group dynamic-constraints" bind:this={panelGroup}>
  <Panel
    default="100px"
    min={dynamicConstraints ? "200px" : "100px"}
    class="panel"
  >
    Panel 1
  </Panel>
  <PanelResizer class="panel-resizer" size="10px" />
  <Panel min="100px" class="panel">Panel 2</Panel>
  <PanelResizer class="panel-resizer" size="10px" />
  <Panel
    min={dynamicConstraints ? "100px" : "400px"}
    max={dynamicConstraints ? "300px" : "700px"}
    class="panel"
  >
    Panel 3
  </Panel>
</PanelGroup>
<button type="button" onclick={toggleDynamicConstraints}>
  Toggle Custom
</button>

<style>
  :global(.dynamic-constraints) {
    width: 1000px !important;
  }

  :global(.panel-group) {
    border: 1px solid rgba(0, 0, 0, 0.3);
    background: rgba(0, 0, 0, 0.1);
    border-radius: 12px;
    box-sizing: border-box;
  }

  :global(.panel) {
    height: 100%;
    width: 100%;
    padding: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
    box-sizing: border-box;
  }

  :global(.panel-resizer) {
    background: red;
  }
</style>
