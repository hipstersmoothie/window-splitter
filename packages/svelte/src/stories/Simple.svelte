<script module>
  import PanelGroup from "../PanelGroup.svelte";
  import Panel from "../Panel.svelte";
  import PanelResizer from "../PanelResizer.svelte";

  let panelGroup;

  /** @type {import("@window-splitter/interface").PanelGroupHandle} */
  export const handle = {
    getId: () => "panel-group",
    getTemplate: () => panelGroup.getTemplate(),
    getState: () => panelGroup.getState(),
    getPixelSizes: () => panelGroup.getPixelSizes(),
    getPercentageSizes: () => panelGroup.getPercentageSizes(),
    setSizes: (sizes) => panelGroup.setSizes(sizes),
  };
</script>

<PanelGroup bind:this={panelGroup} class="panel-group">
  <Panel class="panel">Panel 1</Panel>
  <PanelResizer class="panel-resizer" size="10px" />
  <Panel class="panel" min="100px">Panel 2</Panel>
</PanelGroup>

<style>
  :global(.panel-group) {
    border: 1px solid rgba(0, 0, 0, 0.3);
    background: rgba(0, 0, 0, 0.1);
    border-radius: 12px;
    box-sizing: border-box;
    width: 500px;
  }

  :global(.panel) {
    height: 100%;
    width: 100%;
    padding: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
    box-sizing: border-box;
  }

  :global(.panel-resizer) {
    background: red;
  }
</style>
