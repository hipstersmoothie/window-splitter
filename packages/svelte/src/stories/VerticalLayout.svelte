<script module>
  import PanelGroup from "../lib/PanelGroup.svelte";
  import Panel from "../lib/Panel.svelte";
  import PanelResizer from "../lib/PanelResizer.svelte";

  let panelGroup;

  /** @type {import("@window-splitter/interface").PanelGroupHandle} */
  export const handle = {
    getId: () => "panel-group",
    getTemplate: () => panelGroup.getTemplate(),
    getState: () => panelGroup.getState(),
    getPixelSizes: () => panelGroup.getPixelSizes(),
    getPercentageSizes: () => panelGroup.getPercentageSizes(),
    setSizes: (sizes) => panelGroup.setSizes(sizes),
  };
</script>

<PanelGroup
  bind:this={panelGroup}
  orientation="vertical"
  class="panel-group vertical-example"
>
  <Panel default="30%" min="20%" class="panel">top</Panel>
  <PanelResizer class="panel-resizer" size="10px" />
  <Panel min="20%" class="panel">middle</Panel>
  <PanelResizer class="panel-resizer" size="10px" />
  <Panel default="30%" min="20%" class="panel">bottom</Panel>
</PanelGroup>

<style>
  :global(.vertical-example) {
    height: 322px;
  }

  :global(.panel-group) {
    border: 1px solid rgba(0, 0, 0, 0.3);
    background: rgba(0, 0, 0, 0.1);
    border-radius: 12px;
    box-sizing: border-box;
    width: 500px;
  }

  :global(.panel) {
    height: 100%;
    width: 100%;
    padding: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
    box-sizing: border-box;
  }

  :global(.panel-resizer) {
    background: red;
  }
</style>
