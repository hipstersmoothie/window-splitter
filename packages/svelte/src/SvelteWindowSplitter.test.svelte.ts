import "@testing-library/jest-dom/vitest";

import { cleanup, fireEvent, render, waitFor } from "@testing-library/svelte";
import { afterEach, describe, expect, test, vi } from "vitest";
import * as Cookies from "tiny-cookie";

import Collapsible, {
  handle as collapsibleHandle,
  leftHandle,
  rightHandle,
} from "./stories/Collapsible.svelte";
import AutosaveCookie from "./stories/AutosaveCookie.svelte";
import Autosave, { handle as autosaveHandle } from "./stories/Autosave.svelte";
import AutosaveCollapsible from "./stories/AutosaveCollapsible.svelte";
import DynamicConstraints, {
  handle as dynamicConstraintsHandle,
} from "./stories/DynamicConstraints.svelte";
import Conditional, {
  handle as conditionalHandle,
} from "./stories/Conditional.svelte";
import Simple, { handle as simpleHandle } from "./stories/Simple.svelte";
import VerticalLayout, {
  handle as verticalHandle,
} from "./stories/VerticalLayout.svelte";
import { createTestUtils, dragHandle } from "@window-splitter/interface/test";
import { PanelGroupHandle } from "@window-splitter/interface";

const { expectTemplate, waitForMeasurement, waitForCondition } =
  createTestUtils({
    waitFor,
  });

afterEach(() => {
  cleanup();
});

test("horizontal layout", async () => {
  const component = render(Simple);

  await waitForMeasurement(simpleHandle);

  expect(component.getByText("Panel 1")).toBeInTheDocument();
  expect(component.getByText("Panel 2")).toBeInTheDocument();

  await expectTemplate(simpleHandle, "244px 10px 244px");

  // Should respect the min
  await dragHandle({ delta: 300 });
  await expectTemplate(simpleHandle, "388px 10px 100px");
});

test("vertical layout", async () => {
  const component = render(VerticalLayout);

  await waitForMeasurement(verticalHandle);

  expect(component.getByText("top")).toBeInTheDocument();
  expect(component.getByText("middle")).toBeInTheDocument();
  expect(component.getByText("bottom")).toBeInTheDocument();

  await expectTemplate(verticalHandle, "96px 10px 108px 10px 96px");

  // Should respect the min
  await dragHandle({ delta: 100, orientation: "vertical" });
  await expectTemplate(verticalHandle, "172px 10px 64px 10px 64px");
});

test("Conditional Panels", async () => {
  const component = render(Conditional);

  await waitForMeasurement(conditionalHandle);
  await expectTemplate(conditionalHandle, "244px 10px 244px");

  component.getByText("Expand").click();
  await expectTemplate(conditionalHandle, "244px 10px 134px 10px 100px");

  component.getByText("Close").click();
  await expectTemplate(conditionalHandle, "244px 10px 244px");
});

test("Dynamic constraints", async () => {
  const component = render(DynamicConstraints);

  await waitForMeasurement(dynamicConstraintsHandle);
  await expectTemplate(dynamicConstraintsHandle, "100px 10px 178px 10px 700px");

  component.getByText("Toggle Custom").click();
  await expectTemplate(dynamicConstraintsHandle, "500px 10px 178px 10px 300px");

  component.getByText("Toggle Custom").click();
  await expectTemplate(dynamicConstraintsHandle, "400px 10px 178px 10px 400px");
});

describe("Autosave", () => {
  test("localStorage", async () => {
    localStorage.clear();

    render(Autosave);

    await waitForMeasurement(autosaveHandle);
    await expectTemplate(autosaveHandle, "244px 10px 244px");

    await dragHandle({ delta: 100 });
    await expectTemplate(autosaveHandle, "342px 10px 146px");

    await waitForCondition(() =>
      Boolean(localStorage.getItem("autosave-example-svelte"))
    );
    const obj = JSON.parse(
      localStorage.getItem("autosave-example-svelte") || "{}"
    );
    expect(obj.items).toMatchSnapshot();
  });

  test("callback", async () => {
    localStorage.clear();

    const spy = vi.fn();
    const handle = $state<{ current: PanelGroupHandle | null }>({
      current: null,
    });

    render(AutosaveCollapsible, {
      props: { onCollapseChange: spy, handle },
    });

    await dragHandle({ delta: -200 });
    await expectTemplate(handle.current!, "100px 10px 388px");
    expect(spy).toHaveBeenCalledWith(true);

    cleanup();

    render(AutosaveCollapsible, {
      props: { onCollapseChange: spy, handle },
    });

    await expectTemplate(handle.current!, "100px 10px 388px");
    await dragHandle({ delta: 200 });
    expect(spy).toHaveBeenCalledWith(false);
  });

  test("cookie", async () => {
    // clear cookies
    document.cookie = "test=; expires=Thu, 01 Jan 1970 00:00:00 GMT";

    const handle = $state<{ current: PanelGroupHandle | null }>({
      current: null,
    });

    render(AutosaveCookie, {
      props: { handle, snapshot: undefined },
    });

    await waitForMeasurement(handle.current!);
    await expectTemplate(handle.current!, "245px 10px 245px");

    await dragHandle({ delta: 100 });
    await expectTemplate(handle.current!, "343px 10px 147px");

    await waitForCondition(() =>
      document.cookie.includes("autosave-cookie-svelte")
    );

    expect(document.cookie).toMatchSnapshot();

    const snapshot = Cookies.get("autosave-cookie-svelte");

    cleanup();

    render(AutosaveCookie, {
      props: { snapshot: JSON.parse(snapshot || "{}"), handle },
    });

    await expectTemplate(handle.current!, "343px 10px 147px");
  });
});

describe("imperative panel API", async () => {
  test("panel group", async () => {
    render(Collapsible);

    await waitForMeasurement(collapsibleHandle);

    expect(collapsibleHandle.getPercentageSizes()).toMatchInlineSnapshot(`
      [
        0.5,
        0.020080321285140562,
        0.5,
        0.020080321285140562,
        0.12048192771084337,
      ]
    `);

    expect(collapsibleHandle.getPixelSizes()).toMatchInlineSnapshot(`
      [
        209,
        10,
        209,
        10,
        60,
      ]
    `);

    await expectTemplate(collapsibleHandle, "209px 10px 209px 10px 60px");
  });

  test("panel", async () => {
    render(Collapsible);

    await waitForMeasurement(collapsibleHandle);

    expect(rightHandle.isCollapsed()).toBe(true);
    expect(rightHandle.isExpanded()).toBe(false);

    rightHandle.expand();
    await new Promise((resolve) => setTimeout(resolve, 2000));

    expect(rightHandle.isCollapsed()).toBe(false);
    expect(rightHandle.isExpanded()).toBe(true);

    rightHandle.collapse();
    await new Promise((resolve) => setTimeout(resolve, 2000));

    expect(rightHandle.isCollapsed()).toBe(true);
    expect(rightHandle.isExpanded()).toBe(false);

    // Test the non controlled version

    expect(leftHandle.isCollapsed()).toBe(false);
    expect(leftHandle.isExpanded()).toBe(true);

    leftHandle.collapse();
    await new Promise((resolve) => setTimeout(resolve, 2000));

    expect(leftHandle.isCollapsed()).toBe(true);
    expect(leftHandle.isExpanded()).toBe(false);
    expect(rightHandle.getPercentageSize()).toBe(
      leftHandle.getPercentageSize()
    );
    expect(rightHandle.getPixelSize()).toBe(leftHandle.getPixelSize());

    leftHandle.expand();
    await new Promise((resolve) => setTimeout(resolve, 5000));

    expect(leftHandle.isCollapsed()).toBe(false);
    expect(leftHandle.isExpanded()).toBe(true);
  });
});

test("Keyboard interactions with collapsed panels", async () => {
  render(Collapsible);

  await waitForMeasurement(collapsibleHandle);
  await expectTemplate(collapsibleHandle, "209px 10px 209px 10px 60px");

  const resizer2 = document.getElementById("resizer-2")!;
  fireEvent.keyDown(resizer2, { key: "Enter" });
  await expectTemplate(collapsibleHandle, "209px 10px 169px 10px 100px");

  fireEvent.keyDown(resizer2, { key: "ArrowLeft" });
  fireEvent.keyDown(resizer2, { key: "ArrowLeft" });
  fireEvent.keyDown(resizer2, { key: "ArrowLeft" });
  fireEvent.keyDown(resizer2, { key: "ArrowLeft" });
  await expectTemplate(collapsibleHandle, "209px 10px 165px 10px 104px");

  fireEvent.keyDown(resizer2, { key: "ArrowLeft", shiftKey: true });
  await expectTemplate(
    collapsibleHandle,
    "209.03125px 10px 149.984375px 10px 118.984375px"
  );

  fireEvent.keyDown(resizer2, { key: "Enter" });
  await expectTemplate(
    collapsibleHandle,
    "209.03125px 10px 208.96875px 10px 60px"
  );

  fireEvent.keyDown(resizer2, { key: "Enter" });
  await expectTemplate(
    collapsibleHandle,
    "209.0625px 10px 149.96875px 10px 118.96875px"
  );
});
