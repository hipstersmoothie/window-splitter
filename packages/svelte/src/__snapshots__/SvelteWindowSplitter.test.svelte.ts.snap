// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Autosave > cookie 1`] = `"autosave-cookie-svelte=%7B%22size%22%3A%7B%22x%22%3A0%2C%22y%22%3A0%2C%22width%22%3A500%2C%22height%22%3A198%2C%22top%22%3A0%2C%22right%22%3A500%2C%22bottom%22%3A198%2C%22left%22%3A0%7D%2C%22items%22%3A%5B%7B%22type%22%3A%22panel%22%2C%22currentValue%22%3A%7B%22type%22%3A%22percent%22%2C%22value%22%3A%220.7%22%7D%2C%22min%22%3A%7B%22type%22%3A%22pixel%22%2C%22value%22%3A%220%22%7D%2C%22max%22%3A%221fr%22%2C%22collapsedSize%22%3A%7B%22type%22%3A%22pixel%22%2C%22value%22%3A%220%22%7D%2C%22onCollapseChange%22%3A%7B%7D%2C%22onResize%22%3A%7B%7D%2C%22collapseIsControlled%22%3Afalse%2C%22id%22%3A%221%22%7D%2C%7B%22type%22%3A%22handle%22%2C%22size%22%3A%7B%22type%22%3A%22pixel%22%2C%22value%22%3A%2210%22%7D%2C%22id%22%3A%22resizer%22%7D%2C%7B%22type%22%3A%22panel%22%2C%22currentValue%22%3A%7B%22type%22%3A%22percent%22%2C%22value%22%3A%220.3%22%7D%2C%22min%22%3A%7B%22type%22%3A%22pixel%22%2C%22value%22%3A%220%22%7D%2C%22max%22%3A%221fr%22%2C%22collapsedSize%22%3A%7B%22type%22%3A%22pixel%22%2C%22value%22%3A%220%22%7D%2C%22onCollapseChange%22%3A%7B%7D%2C%22onResize%22%3A%7B%7D%2C%22collapseIsControlled%22%3Afalse%2C%22id%22%3A%222%22%7D%5D%2C%22orientation%22%3A%22horizontal%22%2C%22dragOvershoot%22%3A%220%22%2C%22groupId%22%3A%22autosave-cookie-svelte%22%2C%22autosaveStrategy%22%3A%22cookie%22%2C%22shiftAmount%22%3A15%7D"`;

exports[`Autosave > localStorage 1`] = `
[
  {
    "collapseIsControlled": false,
    "collapsedSize": {
      "type": "pixel",
      "value": "0",
    },
    "currentValue": {
      "type": "percent",
      "value": "0.70081967213114754098",
    },
    "id": "1",
    "max": "1fr",
    "min": {
      "type": "pixel",
      "value": "0",
    },
    "onCollapseChange": {},
    "onResize": {},
    "type": "panel",
  },
  {
    "id": "resizer",
    "size": {
      "type": "pixel",
      "value": "10",
    },
    "type": "handle",
  },
  {
    "collapseIsControlled": false,
    "collapsedSize": {
      "type": "pixel",
      "value": "0",
    },
    "currentValue": {
      "type": "percent",
      "value": "0.29918032786885245902",
    },
    "id": "2",
    "max": "1fr",
    "min": {
      "type": "pixel",
      "value": "0",
    },
    "onCollapseChange": {},
    "onResize": {},
    "type": "panel",
  },
]
`;
